### PR TITLE
p2p/simulation: Test snapshot correctness and minimal benchmark

### PR DESCRIPTION
**Pull request by nolash**
This PR is part one of four PRs that adds test and benchmark to ensure that the snapshot connection content is correct upon creation, and that the correct connections are made after snapshot is loaded. All actual connections should be contained in snapshot, and all connections and only those connections should be in the network after load.

Part one creates the correctness test and adds a vanilla benchmark function.

Part two will move connection methods from swarm/network/simulation to p2p/simulations

Part three will add a snapshot generation binary leveraging the different configurations provided by part two.

Part four will add exhaustive benchmarks with different snapshots generated from part three.

----

This PR also moved the NoopService from the swarm/network/simulation package, since this minimal service is also useful for tests involving p2p/simulations directly.

### CODE REVIEW

File: p2p/simulations/network_test.go [hunk 1]
```diff
@@ -245,6 +245,37 @@ OUTER_TWO:
 			t.Fatalf("Superfluous conn found %v -> %v", ev.Conn.One, ev.Conn.Other)
 		}
 	}
+
+	// This test validates if all connections from the snapshot
+	// are created in the network.
```
Comment (nolash) [inline p2p/simulations/network_test.go]: Actually this description doesn't distinguish it enough from the other, which kind of does this aswell?

We should tell explicitly it's about not return before all is loaded.

File: p2p/simulations/network_test.go [hunk 2]
```diff
@@ -18,14 +18,268 @@ package simulations
...
+		},
+	})
+
+	// create network
+	network := NewNetwork(adapter, &NetworkConfig{
+		DefaultService: "noopwoop",
+	})
+	// \todo consider making a member of network, set to true threadsafe when shutdown
+	runningOne := true
+	defer func() {
```
Comment (justelad) [inline p2p/simulations/network_test.go]: i'm not sure i understand why do you actually need this.
you're setting:
```
	// shut down sim network
	runningOne = false
	sub.Unsubscribe()
	network.Shutdown()
```
and the snapshot network is being explicitly shut down later in the test.

File: p2p/simulations/network_test.go [hunk 3]
```diff
@@ -18,14 +18,268 @@ package simulations
...
+	runningOne = false
+	sub.Unsubscribe()
+	network.Shutdown()
+
+	// check that we have all the expected connections in the snapshot
+	for nodid, nodConns := range checkIds {
+		for _, nodConn := range nodConns {
+			var match bool
+			for _, snapConn := range snap.Conns {
+				if snapConn.One == nodid && snapConn.Other == nodConn {
```
Comment (justelad) [inline p2p/simulations/network_test.go]: you can cram these two `if`'s into one:
`if s.One == n1 && s.Other == n2 || s.One == n2 && s.Other == n1 ...`

File: p2p/simulations/network_test.go [hunk 4]
```diff
@@ -18,14 +18,268 @@ package simulations
...
+			}
+		}
+	}
+
+	// check that we have all expected connections in the network
+	for _, snapConn := range snap.Conns {
+		var match bool
+		for nodid, nodConns := range checkIds {
+			for _, nodConn := range nodConns {
+				if snapConn.One == nodid && snapConn.Other == nodConn {
```
Comment (justelad) [inline p2p/simulations/network_test.go]: can be merged into one `if`

**Pull request merged**